nbinteract/plotting.py

```diff
@@ -585,25 +599,30 @@ def call_params(component, opts):
             trait: maybe_call(val, opts)
             for trait, val in component.items()
         }
-
-    params = _merge_with_defaults(params)
-
-    x_sc = x_sc(**call_params(params['x_sc'], options))
-    y_sc = y_sc(**call_params(params['y_sc'], options))
-    options = {**options, **{'x_sc': x_sc, 'y_sc': y_sc}}
-
-    x_ax = x_ax(**call_params(params['x_ax'], options))
-    y_ax = y_ax(**call_params(params['y_ax'], options))
-    options = {**options, **{'x_ax': x_ax, 'y_ax': y_ax}}
-
-    marks = [
-        mark_cls(**call_params(mark_params, options))
-        for mark_cls, mark_params in zip(marks, params['marks'])
-    ]
-    options = {**options, **{'marks': marks}}
-
-    fig = fig(**call_params(params['fig'], options))
-
+    if callable(fig):
+        params = _merge_with_defaults(params)
+
+        x_sc = x_sc(**call_params(params['x_sc'], options))
+        y_sc = y_sc(**call_params(params['y_sc'], options))
+        options = {**options, **{'x_sc': x_sc, 'y_sc': y_sc}}
+
+        x_ax = x_ax(**call_params(params['x_ax'], options))
+        y_ax = y_ax(**call_params(params['y_ax'], options))
+        options = {**options, **{'x_ax': x_ax, 'y_ax': y_ax}}
+        marks = []
+        options = {**options, **{'marks': marks}}
+        fig = fig(**call_params(params['fig'], options))
+    else:
+        params = _merge_with_defaults(params)
+        x_sc = fig.axes[0].scale
+        y_sc = fig.axes[1].scale
+        options = {**options, **{'x_sc': x_sc, 'y_sc': y_sc}}
+        marks = [
+            mark_cls(**call_params(mark_params, options))
+            for mark_cls, mark_params in zip(marks, params['marks'])
+                ]
+        options = {**options, **{'marks': marks}}
+        fig.marks = fig.marks+marks
```

Comment:
The issue with the function above is that it does a lot of stuff. It branches on whether the figure is a class or an object. If it's a class, we don't create marks. If it's an object, we pull scales from the existing figure and create marks using those scales.

Long `if`/`else` statements suggest that we can split up this method into two pieces. In this case, it looks like we can split this code into a `_create_marks` method and a `_create_fig` method.

Then we can do something like

```python
def scatter(...):
    ...
    fig = fig or _create_fig(...)
    [scat] = _create_marks(fig=fig, ...)
    fig._add_marks([scat])
    ...
```

which should make the individual methods a lot easier to write.

Reply:
Split into _create_fig(...), _create_marks(fig=fig, marks=marks), _add_marks([lst_of_marks])